lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js

```diff
@@ -238,5 +241,157 @@
+                        __chakraLibrary.raiseLengthIsTooBig("Array.prototype.flatMap");
+                    }
+                    //  2. Perform ? CreateDataPropertyOrThrow(target, ! ToString(targetIndex), element).
+                    __chakraLibrary.arrayCreateDataPropertyOrThrow(target, targetIndex, element);
+                    //  3. Increase targetIndex by 1.
+                    ++targetIndex;
+                }
+            }
+            //  d. Increase sourceIndex by 1.
+            ++sourceIndex;
+        }
+        //4. Return targetIndex.
+        return targetIndex;
+    });
+
+    platform.registerFunction(FunctionsEnum.ArrayFlat, function (depth) {
+        "use strict";
+        // this version excludes the mapper function
+        //1. Let O be ? ToObject(this value).
+        //2. Let sourceLen be ? ToLength(? Get(O, "length")).
+        let o, sourceLen;
+        
+        if (__chakraLibrary.isArray(this)) {
+            o = this;
+            sourceLen = o.length;
+        } else {
+            if (this === null || this === undefined) {
+                __chakraLibrary.raiseThis_NullOrUndefined("Array.prototype.flat");
+            }
+            o = __chakraLibrary.Object(this);
+            sourceLen =  __chakraLibrary.GetLength(o);
+        }
+        //3. Let depthNum be 1.
+        //4. If depth is not undefined, then
+        //5. Set depthNum to ? ToInteger(depth).
+        const depthNum = depth !== undefined ? __chakraLibrary.toInteger(depth) : 1;
+        //6. Let A be ? ArraySpeciesCreate(O, 0).
+        const A = __chakraLibrary.arraySpeciesCreate(o, 0);
+        //7. Perform ? FlattenIntoArray(A, O, sourceLen, 0, depthNum).
+        __chakraLibrary.FlattenIntoArray(A, o, sourceLen, 0, depthNum);
+        //8. Return A.
+        return A;
+    });
+
+    platform.registerFunction(FunctionsEnum.ArrayFlatMap, function (mapperFunction, thisArg) {
+        "use strict";
+        //1. Let O be ? ToObject(this value).
+        //2. Let sourceLen be ? ToLength(? Get(O, "length")).
+        let o, sourceLen;
+        if (__chakraLibrary.isArray(this)) {
+            o = this;
+            sourceLen = o.length;
+        } else {
+            if (this === null || this === undefined) {
+                __chakraLibrary.raiseThis_NullOrUndefined("Array.prototype.flatMap");
+            }
+            o = __chakraLibrary.Object(this);
+            sourceLen =  __chakraLibrary.GetLength(o);
+        }
+        //3. If IsCallable(mapperFunction) is false throw a TypeError exception
+        if (!__chakraLibrary.isCallable(mapperFunction)) {
+            __chakraLibrary.raiseFunctionArgument_NeedFunction("Array.prototype.flatMap");
+        }
+        //4. If thisArg is present, let T be thisArg; else let T be undefined
+        //5. Let A be ? ArraySpeciesCreate(O, 0).
+        const A = __chakraLibrary.arraySpeciesCreate(o, 0);
+        //6. Perform ? FlattenIntoArray(A, O, sourceLen, 0, depthNum).
+        if (!thisArg) {
```

Comment:
I think this is the reason the test is failing; you are just checking the truthiness. Undefined works because when you call a strict function `foo()` it will get `undefined` as its this, but with this check calling passing false or 0 or the empty string or any other falsy value would fail I'd suggest changing this to `thisArg === undefined`, and in all other cases going with the bind in the else.

Reply:
Updated